docs: firmware updates through a fwupd container (#70)
kinvolk/racker: docs/usage/firmware.md

---
title: Firmware Updates
weight: 70
---

Flatcar OS updates include the Linux firmware package which contains firmware binaries for those firmwares that get loaded at runtime.
Some devices, however, need their firmware updates flashed out of band and may even require reboots for the updates to take effect.
The vendors of those devices supply a firmware binary and instructions how to flash it.

The firmware updates on a Kubernetes cluster can be done with a special firmware update container that runs on each node.
This container can either be run manually via SSH or through a Kubernetes _DaemonSet_.

Since the firmware updates are vendor-specific, there are no generic instructions on how to do them.

> To improve this situation and provide a unified and automated experience, the _Linux Vendor Firmware Service_ was started.

Reviewer: ```suggestion
To improve this situation and provide a unified and automated experience, the [Linux Vendor Firmware Service](https://fwupd.org/) was started.
```

[Install]\n\
WantedBy=multi-user.target\n\
'  > /etc/systemd/system/fwupd-action.service
RUN systemctl enable fwupd-action.service
RUN echo OverrideESPMountPoint=/boot >> /etc/fwupd/uefi_capsule.conf
# Work around a problem with a missing dbus.service unit file
RUN mkdir -p /etc/systemd/system && ln -fs /usr/lib/systemd/system/dbus-broker.service /etc/systemd/system/dbus.service
# Mask udev instead of making /sys/ read-only as suggested in https://systemd.io/CONTAINER_INTERFACE/
RUN systemctl mask systemd-udevd.service systemd-modules-load.service systemd-udevd-control.socket systemd-udevd-kernel.socket
ENV container=docker
VOLUME [ "/sys/fs/cgroup" ]
# Set up a /dev/console link to have the same behavior with or without -it
CMD [ "/bin/sh", "-c", "if ! [ -e /dev/console ] ; then socat -u pty,link=/dev/console stdout & fi ; exec /sbin/init" ]
# TODO: maybe need to set up /usr/libexec/fwupd/efi/fwupdx64.efi(.signed) and DisableShimForSecureBoot?
```

Now build a container image from it and push it to a registry of your choice:

```
docker build -t my-registry/fwupd:latest .
docker push my-registry/fwupd:latest
```

> Then, on each node, run it with a Kubernetes _DaemonSet_ as a one-time action or with the following Docker command via SSH:

Reviewer: ```suggestion
Then, on each node, run it with the following Docker command via SSH:
```

I think the phrasing was a bit confusing because it talked about the DaemonSet first, but it referred to the latter example instead.

```
docker run --privileged --net host --rm -v /sys/fs/cgroup:/sys/fs/cgroup:ro -v /sys/firmware/efi/efivars:/sys/firmware/efi/efivars -v /boot:/boot my-registry/fwupd:latest
```

> You can also use a Kubernetes _DaemonSet_ but keep in mind that the container exits directly after finishing its action and that you need to use the equivalent privileges as in the command above.

Reviewer: ```suggestion
You can also use a Kubernetes [DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/) but keep in mind that the container exits directly after finishing its action and that you need to use the equivalent privileges as in the command above.
```

---
title: Firmware Updates
weight: 70
---

> Flatcar OS updates include the Linux firmware package which contains firmware binaries for those firmwares that get loaded at runtime.

Reviewer 2: Shouldnt this a full name as we strive to use it everywhere ?
```suggestion
Flatcar Container Linux OS updates include the Linux firmware package which contains firmware binaries for those firmwares that get loaded at runtime.
```

> Flatcar OS updates include the Linux firmware package which contains firmware binaries for those firmwares that get loaded at runtime.

Reviewer 2: ```suggestion
Flatcar OS updates include the Linux firmware package which contains firmware binaries for  firmware that gets loaded at runtime.
```

Author: thanks

Some devices, however, need their firmware updates flashed out of band and may even require reboots for the updates to take effect.
The vendors of those devices supply a firmware binary and instructions how to flash it.

For example, the update utility may be a static binary and a specific firmware file for your server.
These would need to be downloaded to the node and directly executed there through SSH.
One could also automate the firmware updates a bit more on a Kubernetes cluster with a special firmware update container that runs on each node.

Since the firmware updates are vendor-specific, there are no generic instructions on how to do them.
To improve this situation and provide a unified and automated experience, the [Linux Vendor Firmware Service](https://fwupd.org/) was started.

> The system service `fwupd` can fetch firmware updates and flash them for supported devices.

Reviewer 2: ```suggestion
The system service `fwupd` can fetch firmware updates and flash them on supported devices.
```

Reviewer 2: maybe I'm wrong with the usage of `on`

Author: my idea was to use _for_ for fetch+flash

> Not all vendors and not all devices are part yet of this effort but it is the only generic mechanism that can be documented as an example.

Reviewer 2: ```suggestion
Not all vendors or devices are part yet of this effort but it is the only generic mechanism that can be documented as an example.
```

> ## Firmware Update Container with fwupd

Reviewer 2: This would be a lot more easier to follow for the user if the instructions were mentioned as steps.

Author: done
